### PR TITLE
Introduce Argon2 hashing algorithm

### PR DESCRIPTION
**Pull request by technion**
Hi,

I have added the newer Argon2 algorithm as an option for hashing passwords. This will not pass CI unfortunately due to the current Travis situation.

Please ensure your pull request includes the following:

- [x] Description of changes
- [x] Update to CHANGELOG.md with short description and link to pull request
- [x] Changes have related RSpec tests that ensure functionality does not break

### CODE REVIEW

**Comment by jojoa** [comment]
Nice Idea.

**Comment by athix** [comment]
@technion I pushed up a commit to add Github actions to master. Assuming it passes, can you merge master into your PR? I'll try to take a look at this and merge it when I have some time.

**Comment by technion** [comment]
@athix Thanks for that. I know moving my own projects from Travis to GA was a huge pain.

**Review by athix — CHANGES_REQUESTED** [review]
I'll be doing a deep dive on the hashing algorithms in the near future as a part of introducing Argon2 as a new default to the v1 rework. I'll hold off on merging this until I've had the time to familiarize myself with them thoroughly.

Until then, I'm curious why `cost_matches?` was left unimplemented. The BCrypt implementation supports it, so I would imagine the newer Argon2 should also be capable of doing so; unless there was some security concern discovered with it?

File: lib/sorcery/crypto_providers/argon2.rb
Comment (technion) [reply lib/sorcery/crypto_providers/argon2.rb:36]: I think your answer is that it's worth implementing. It's not defined in the Argon2 gem, but it's plausible with the Argon2 format. I'll get this done shortly.

Comment (athix) [reply lib/sorcery/crypto_providers/argon2.rb:36]: Weird that the Argon2 gem doesn't support this. I ran into it when looking at the structure of its password class vs BCrypt's. Honestly, this seems like something that should live upstream in the gem, like it does with BCrypt. I'll see if the Argon2 maintainer is open to that change when I get to Argon2 in the rework (should be in the next week, at my current pace).

Comment (technion) [reply lib/sorcery/crypto_providers/argon2.rb:36]: The upstream maintainer is me. The gem was based as much as possible on the upstream C library, and aligning with the way it's used was a bigger priority than acting like a Bcrypy gem. That said, this is a very trivial one liner to implement, I'm just testing it.

Comment (athix) [reply lib/sorcery/crypto_providers/argon2.rb:36]: > The upstream maintainer is me.

Oh... I feel dumb for not noticing that sooner. :sweat_smile: I've been somewhat bullish with my attention while I plunk away on the Sorcery v1 rework.

> The gem was based as much as possible on the upstream C library, and aligning with the way it's used was a bigger priority than acting like a BCrypt gem. That said, this is a very trivial one liner to implement, I'm just testing it.

Would you be okay with deviating slightly from the 'unaltered' design tenant for the ruby wrapper? I feel it would save a lot of downstream developer effort if we exposed some key helpers within the ruby wrapper, rather than forcing those helpers to be downstream.

This would of course still be a git sub-module, and keep all of the functional aspects delegated to the C implementation. I would also be willing to help co-maintain the wrapper, and open a PR to add these changes where we can work together to find a good balance between maintainability and downstream benefits.

Here's my work on the BCrypt implementation last night, it should hopefully give you an idea of the helpers that I'm looking for: [bcrypt.rb](https://github.com/Sorcery/sorcery-rework/blob/e6e78b71226ba862b9ce167e66306eb6f8ba4fd1/sorcery-core/lib/sorcery/crypto_providers/bcrypt.rb)

**Comment by technion** [comment]
I just pushed a trivial update that helps out here.
Just to complicate things, I'm using 'm' as the 'cost', since it's the parameter most would recommend adjusting. However, Argon2 costs actually have three tunable parameters, and I'll be working on exposing the third in https://github.com/technion/ruby-argon2/issues/42.

I'm obviously expecting Sorcery can't easily take all these on without impacting the API, and that's probably OK for the normal person, but it's worth mentioning. 

I have no objections to pushing helpers into the gem when they're seen as helpers people might want. I do feel the root of this discussion is that I followed more closely BCrypt as implemented in most other languages.

For example in PHP:

    password_verify ( string $password , string $hash ) : bool

Or Javascript:

    const match = await bcrypt.compare(password, user.passwordHash);

I feel that's much less trouble prone than Ruby's BCrypt::Password.new(digest) followed by an equality check with a magically overridden ==, although I also appreciate it makes some of these API helpers more necessary.

**Comment by athix** [comment]
> I feel that's much less trouble prone than Ruby's BCrypt::Password.new(digest) followed by an equality check with a magically overridden ==, although I also appreciate it makes some of these API helpers more necessary.

I actually got bit by the weirdness of how they handle `==` in the BCrypt ruby library just earlier... RSpec `eq` is working when comparing a string and a BCrypt password object, but for whatever reason that same comparison fails when I used it for `digest_matches?(digest, password)`. No idea what witchcraft is hidden in there to cause that kind of behavior...

I'm in the middle of writing the Argon2 Sorcery implementation in the rework, but I'll have time today to get started on adding those helpers. Would you be interested in chatting via Discord/similar to hammer out the details?

**Comment by athix** [comment]
Speaking of that `==` bug, I figured it out and pushed up a potential fix as https://github.com/codahale/bcrypt-ruby/pull/245. Turns out `==` will always hash your input, even if the input is a valid bcrypt hash already. This explains why coercing the BCrypt::Password object to a String, then doing your comparison will work, as well as why doing `bcrypt_object == original_secret` also works, but doing `bcrypt_object == identical_hash` does not.

**Comment by athix** [comment]
> I'm obviously expecting Sorcery can't easily take all these on without impacting the API, and that's probably OK for the normal person, but it's worth mentioning.

@technion the rework is already in "needs a major migration" territory, so I think it's worth abstracting the crypto interface to better support the various options for crypto providers. There's the 3 costs for Argon2, but it might also be possible that some newer algo comes out with other params too, or we add support for scrypt/other modern crypto providers.

I'll look into following a similar methodology as for what's being used in plugins to extend the crypto config. I'd imagine it'll take me an afternoon, but should be quite doable, and allow precise editing of the Argon2 costs per application (or even per auth flow with how the new config supports multiple flows within a single app. You could for example, have higher costs associated with your admin login flow... If that would be desirable anyway.)

**Comment by athix** [comment]
@technion unfortunately with how the old version of Sorcery currently works, to support this we would need to ship both BCrypt and Argon2 as a dependency for all apps. (see [sorcery.rb](https://github.com/Sorcery/sorcery/blob/master/lib/sorcery.rb) and [sorcery.gemspec](https://github.com/Sorcery/sorcery/blob/master/sorcery.gemspec#L36)) Considering this, I'm going to leave adding Argon2 as purely a Sorcery v1 feature, where devs won't be forced to bundle libraries that they aren't using.

Argon2 support is already present in the rework, so the only blocker is finishing the v1 rework and updating the docs / publishing the migration guide.

**Pull request closed without merging**
